### PR TITLE
Documentation for dapper.tools.math and bug fixes for dapper.tools.series

### PR DESCRIPTION
**Pull request by yumengch**
Hi Patrick

Here are several updates on dapper:
1. Documentation update for dapper.tools.math. I adopt the numpy style docstring here and I need to note that: all functions decorated by @np.vectorize are not documented automatically, which seemts to be a feature in pdoc.
2. I fixed some bugs we discussed in repository issues
3. I added the Vissio-Lucarini model into DAPPER, which I will take more care of later.
4. I added two functions for adjoint model and tangent linear model in dapper.tools.math using RK4 and the integration of these methods to the package is not very natural at the moment.

### CODE REVIEW

**Comment by patricknraanes** [comment]
Hi Yumeng!

Thanks a lot for this. It's great to see all the work you've been doing, and it's really good. Especially from your documentation additions I can see that you have a good grasp of the code!

Unfortunately, I don't know how to merge this at the moment. The reason is of course that there has been a lot of development from my side on DAPPER since Dec 14. And much of it pertains to the same things you have been doing: documentation. This means that a there are a lot of lines will have conflicts with this PR. Also, for example, the math module does not exist anymore. Several functions have been removed, and the others have gotten improved documentation (though not as good as some of what you put in), and moved elsewhere.

So, in order to merge, I think we need to (close this PR and) cherry-pick the commits one-by-one, which probably will involve a lot of manual editing. If you wish, I can do it. But then the contributions would not appear as your (re-newed) PR, which would be a shame considering the work you've put in. Please let if know if you prefer I take should care of it, or if you will.

I'm also quite new at github collaboration, so I thank you for your patience if I seem irresponsive. However, I believe that our collaboration process will get smoother with time. This will happen naturally as we learn, but also because the project is getting more mature, and so have less need of wholesale changes. In order to minimise future conflicts, I suggest that we try to push a little more frequently (and that you pull even more frequently, especially before submitting PRs), and possibly plan & agree beforehand what we will be working on.

Replies to your points:

1. Great that you follow the numpy documentation (as I have started to (try to) do)! Some remarks:
   1. I discovered this problem of `np.vectorize`, and addressed it by making [np_vectorize](https://github.com/nansencenter/DAPPER/search?l=Python&q=np_vectorize)
   2. Double backticks are mostly a remnant from when the documentation was written in rst. Now, though, I adapt the use of single backticks because the docs are mostly in markdown, so I don't think there is a point to double backticks. I say *mostly* because some rst features still seem to work for pdoc3, like `.. note::`. So I'm not sure exactly how pdoc3 decides to interpret things. Do you? Anyhow, please let me know if I'm missing something.
   3. I'm also (from now on) using the convention that docstrings begin on the same line as the opening quotation marks. This is a little more compact. Do you have a preference for beginning on the next line instead?

2. As we've discussed in their resepective issues, these were very useful, thank you!

3. I'm exited about the new model, and look forward to seeing more of it.

4. Could you clarify for me what `RK4_adj` and `RK4_linear` do compared to `integrate_TLM`?
   For example, note that this already enables running the extended KF
   and [4D-Var](https://github.com/nansencenter/DAPPER/search?l=Python&q=var4d)

Some other notes:

- The development toolchain has been significantly improved, so you might want to re-initialize your environment. Some of the changes are: 
  - A good deal of code has been shifted to the external packages `mpl-tools`, `patlib`, and `struct-tools`
  - Doctests are now integrated. Please add them whenever you see fit. Use `# doctest: +SKIP` to not run them for a particular example. Please run all tests with the single command `pytest` before any PR.
  - There is an official linter: flakehell. This is a version of flake8 that allows configuration within `pyproject.toml`, where the tests are also configured. Linting checks are included with the pre-commit hooks.

**Comment by yumengch** [comment]
Hi Patrick

Thanks for your comments and you did a lot of development!

I agree that things will get smoother after we gradually get used to this process and I seeing your commits in GitHub more frequently.

> So, in order to merge, I think we need to (close this PR and) cherry-pick the commits one-by-one, which probably will involve a lot of manual editing. If you wish, I can do it. But then the contributions would not appear as your (re-newed) PR, which would be a shame considering the work you've put in. Please let if know if you prefer I take care of it, or if you will.

Since this is the main repository, I believe it is reasonable to close this PR and I will push for a new one after I merged your new version. I think the future work will require a more frequent commits into GitHub at least in the dev1 branch and I will catch up with those new updates. 

> - Double backticks are mostly a remnant from when the documentation was written in rst. Now, though, I adapt the use of single backticks because the docs are mostly in markdown, so I don't think there is a point to double backticks. I say mostly because some rst features still seem to work for pdoc3, like .. note::. So I'm not sure exactly how pdoc3 decides to interpret things. Do > you? Anyhow, please let me know if I'm missing something.
> - I'm also (from now on) using the convention that docstrings begin on the same line as the opening quotation marks. This is a little more compact. Do you have a preference for beginning on the next line instead?

I will follow this convention and we can always adapt if anything doesn't work.

> Could you clarify for me what RK4_adj and RK4_linear do compared to integrate_TLM?
> For example, note that this already enables running the extended KF and 4D-Var

This is something I haven't completely integrate into DAPPER. The intention was to use it for some of my toy code for a weak constraint 4DVar. I put them into the DAPPER code since I think they might be useful in the future.

The code is translated from an old Matlab code for 4D-Var. The `RK4_linear` uses a time-dependent tangent linear model and I think `integrate_TLM` assumes TLM is constant. 

RK4_adj is used to compute `M.T@x` where `M.T` is the adjoint model and `x` is a vector. This form appears in the gradient of 4D-Var and this function always uses vectors and can avoids the computation of 2-D matrix TLM and transpose it afterwards. 

Currently both functions require independent python functions in the model, which I only implemented for L96 at the moment.

I'm happy for the new dev toolchain and I will check them!

**Comment by patricknraanes** [comment]
> The code is translated from an old Matlab code for 4D-Var. The `RK4_linear` uses a time-dependent tangent linear model and I think `integrate_TLM` assumes TLM is constant.

Interesting. Time-dependence would be nice. It should be accompanied by benchmarks (L96/Var4D?) comparing to non-time-dependent method.

> RK4_adj is used to compute `M.T@x` where `M.T` is the adjoint model and `x` is a vector. This form appears in the gradient of 4D-Var and this function always uses vectors and can avoids the computation of 2-D matrix TLM and transpose it afterwards.

This is also good, but IMO has lower priority because dapper doesn't focus on big models, making it less necessary.

It would also be best if both features were offered as options to `integrate_TLM` somehow, in order to avoid duplication.

**Comment by patricknraanes** [comment]
I see there have been several new commits. Does it mean that this PR has been updated, and that I might now be able to merge?

**Comment by yumengch** [comment]
> I see there have been several new commits. Does it mean that this PR has been updated, and that I might now be able to merge?

I removed the RK4_linear and RK4_adj at the moment. I think it's safe to merge. I will add a bit more for VL model (e.g. demo.py etc.) in the future but the code itself is safe to use.

**Comment by yumengch** [comment]
> > The code is translated from an old Matlab code for 4D-Var. The `RK4_linear` uses a time-dependent tangent linear model and I think `integrate_TLM` assumes TLM is constant.
> 
> Interesting. Time-dependence would be nice. It should be accompanied by benchmarks (L96/Var4D?) comparing to non-time-dependent method.
> 
> > RK4_adj is used to compute `M.T@x` where `M.T` is the adjoint model and `x` is a vector. This form appears in the gradient of 4D-Var and this function always uses vectors and can avoids the computation of 2-D matrix TLM and transpose it afterwards.
> 
> This is also good, but IMO has lower priority because dapper doesn't focus on big models, making it less necessary.
> 
> It would also be best if both features were offered as options to `integrate_TLM` somehow, in order to avoid duplication.

I will think a bit about it and do it later.

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:147]:
```suggestion
    """Specify identity observations of a subset of obs. indices.
    
```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:153]:
```suggestion
        Length of state vector
```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:178]:
```suggestion

    It is not a function of time.
```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:183]:
```suggestion
        Length of state vector
```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:195]:
```suggestion
    """Provide a range of `Ny` equispaced integers between `0` and `Nx-1`.

```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:204]:
```suggestion
        Determines if `Nx == 0`.
```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:209]:
```suggestion
    integers: ndarray
        The list of integers.
```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:210]:
```suggestion

```

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:220]:
```suggestion
    array([0, 2, 4, 6, 8])
```

File: dapper/stats.py
Comment (patricknraanes) [inline dapper/stats.py:688]:
```suggestion
    """Like `center`, but only return the anomalies (not the mean)."""
```

File: dapper/stats.py
Comment (patricknraanes) [inline dapper/stats.py:724]:
```suggestion
        Importance weights. Must sum to 1.
```

File: dapper/stats.py
Comment (patricknraanes) [inline dapper/stats.py:758]:
```suggestion
    Reference
```

File: dapper/stats.py
Comment (patricknraanes) [inline dapper/stats.py:743]:
```suggestion
        Importance weights. Must sum to 1.
```

File: dapper/stats.py
Comment (patricknraanes) [inline dapper/stats.py:748]:
```suggestion
        The "effective" size of the weighted ensemble.
        If not provided, it is computed from the weights.
        The unbiasing factor is $$ N_{eff} / (N_{eff} - 1) $$.
```

File: dapper/xp_launch.py
Comment (patricknraanes) [inline dapper/xp_launch.py:467]: I'm not really in favor of this, because I find the original format easier to read for humans. Any reason you suggested this change?

**Review by patricknraanes — COMMENTED** [review]
Ok, I reviewed the PR. All good. I made minor changes. I think you need to approve or comment on them.

File: dapper/xp_launch.py
Comment (yumengch) [reply dapper/xp_launch.py:467]: I nearly forgot this change. I did this in an ad-hoc way as I'm running DAPPER in Windows (at least generating files in Windows) and the system does not allow colon in filenames.  It seems most Windows files with timestamp use %H%M%S without any symbols in between. I think we should either keep the original format or add one condition for Windows system, or simply delete the colons.

Comment (patricknraanes) [reply dapper/xp_launch.py:467]: Very good point! Windows should be supported, but I'm running on Mac and Linux, and so have a hard time testing with windows. I had this bug once: it seems in Windows you cannot have folders named "utils", which was funny.

I really want human-readability, and personally found it very hard to read without separators. I suggest we stick with your change, using dashes as separators.

BTW, others have had the same [question](https://old.reddit.com/r/ISO8601/comments/d6mdaz/iso_8601_compliant_filenames_on_windows/).

Comment (yumengch) [reply dapper/xp_launch.py:467]: Good, I will keep the change. I just tested and I think the quotation marks are the problem for "utils". It is always interesting to find out this kind of design differences.

**Comment by yumengch** [comment]
> Ok, I reviewed the PR. All good. I made minor changes. I think you need to approve or comment on them.
> 
> Edit: Or maybe it is I who must approve? Do you have the option?

I just see the edit. I cannot merge it I think and I guess you are the only one who can decide if you want to merge it.

**Comment by patricknraanes** [comment]
Are you sure you did not resolved them? See screenshot:
res

I'm trying to be sure I understand everything pressing the "merge" button.

**Comment by yumengch** [comment]
I pressed the resolved button as I think it means I agree with the changes you made.

**Review by patricknraanes — APPROVED** [review]

File: dapper/mods/utils.py
Comment (patricknraanes) [inline dapper/mods/utils.py:223]:
```suggestion
```

**Review by patricknraanes — APPROVED** [review]

**Pull request merged**
